osf/migrations/__init__.py

```diff
@@ -119,11 +120,19 @@ def update_admin_permissions(verbosity=0):
 def update_provider_auth_groups(verbosity=0):
     # TODO: determine efficient way to only do this if perms change
     from osf.models.provider import AbstractProvider
+    from django.db import transaction
     for subclass in AbstractProvider.__subclasses__():
-        for obj in subclass.objects.all():
-            obj.update_group_permissions()
-            if verbosity > 0:
-                logger.info('Updated perms for {} {}'.format(obj.type, obj._id))
+        # The exception handling here allows us to make model changes to providers while also checking their permissions
```

Comment:
Could you be a bit more explicit with what's happening here? I'm not super-familiar with this code, and I'm not sure on inspection why a ProgrammingError would come up in this instance.

Reply:
I'm not entirely sure myself why we are updating permissions prior to completing all migrations, but according to @mfraezz it is necessary. I think the real root of this problem is that the `post_migrate` ready signal is acting up and not waiting for the field to be added to Provider before modifying it.

Reply:
The function itself is necessary for reviews groups to 1) exist and 2) have the correct set of permissions (see [perms](https://github.com/CenterForOpenScience/osf.io/blob/develop/osf/utils/permissions.py#L24-L53)) in the event that perms change (they have a couple times, but not recently).

As to why a `ProgrammingError` was being raised here -- an [earlier migration](https://github.com/CenterForOpenScience/osf.io/blob/develop/osf/migrations/0136_preprint_node_divorce.py#L94) in the tree is manually emitting the `post_migrate` signal. An underlying assumption of that signal appears to be "the model schemas in database and models in code are in sync," which does not hold true in this case, causing the queries in this function to look for columns that don't exist.

The other option for fulfilling the need for this function would be resolving the `TODO` at the top; determine a way to do this only when perms change (that doesn't require someone remembering that it needs to happen).

Reply:
Cool, thank you.